beacon-chain/state/stategen/replayer.go

```diff
@@ -0,0 +1,347 @@
+	defer span.End()
+	currentSlot := r.cs.CurrentSlot()
+	if target > currentSlot {
+		return nil, nil, errors.Wrap(ErrFutureSlotRequested, fmt.Sprintf("requested=%d, current=%d", target, currentSlot))
+	}
+	_, b, err := r.canonicalBlockForSlot(ctx, target)
+	if err != nil {
+		return nil, nil, errors.Wrap(err, fmt.Sprintf("unable to find replay data for slot=%d", target))
+	}
+	s, descendants, err := r.ancestorChain(ctx, b)
+	if err != nil {
+		return nil, nil, errors.Wrap(err, "failed to query for ancestor and descendant blocks")
+	}
+
+	return s, descendants, nil
+}
+
+// canonicalBlockForSlot uses HighestSlotBlocksBelow(target+1) and the CanonicalChecker
+// to find the highest canonical block available to replay to the given slot.
+func (r *canonicalChainer) canonicalBlockForSlot(ctx context.Context, target types.Slot) ([32]byte, block.SignedBeaconBlock, error) {
+	for target > 0 {
+		hbs, err := r.h.HighestSlotBlocksBelow(ctx, target+1)
+		if err != nil {
+			return [32]byte{}, nil, errors.Wrap(err, fmt.Sprintf("error finding highest block w/ slot <= %d", target))
+		}
+		if len(hbs) == 0 {
+			return [32]byte{}, nil, errors.Wrap(ErrNoBlocksBelowSlot, fmt.Sprintf("slot=%d", target))
+		}
+		r, b, err := r.bestForSlot(ctx, hbs)
+		if err == nil {
+			// we found a valid, canonical block!
+			return r, b, nil
+		}
+
+		// we found a block, but it wasn't considered canonical - keep looking
+		if errors.Is(err, ErrNoCanonicalBlockForSlot) {
+			// break once we've seen slot 0 (and prevent underflow)
+			if hbs[0].Block().Slot() == 0 {
+				break
+			}
+			target = hbs[0].Block().Slot() - 1
+			continue
+		}
+		return [32]byte{}, nil, err
+	}
+	b, err := r.h.GenesisBlock(ctx)
+	if err != nil {
+		return [32]byte{}, nil, errors.Wrap(err, "db error while retrieving genesis block")
+	}
+	root, _, err := r.bestForSlot(ctx, []block.SignedBeaconBlock{b})
+	if err != nil {
+		return [32]byte{}, nil, errors.Wrap(err, "problem retrieving genesis block")
+	}
+	return root, b, nil
+}
+
+// bestForSlot encapsulates several messy realities of the underlying db code, looping through multiple blocks,
+// performing null/validity checks, and using CanonicalChecker to only pick canonical blocks.
+func (r *canonicalChainer) bestForSlot(ctx context.Context, hbs []block.SignedBeaconBlock) ([32]byte, block.SignedBeaconBlock, error) {
+	for _, b := range hbs {
+		if helpers.BeaconBlockIsNil(b) != nil {
+			continue
+		}
+		root, err := b.Block().HashTreeRoot()
+		if err != nil {
+			// use this error message to wrap a sentinel error for error type matching
+			wrapped := errors.Wrap(ErrInvalidDBBlock, err.Error())
+			msg := fmt.Sprintf("could not compute hash_tree_root for block @ slot=%d", b.Block().Slot())
+			return [32]byte{}, nil, errors.Wrap(wrapped, msg)
+		}
+		canon, err := r.c.IsCanonical(ctx, root)
+		if err != nil {
+			return [32]byte{}, nil, errors.Wrap(err, "error from blockchain info fetcher IsCanonical check")
```

Comment:
```suggestion
			return [32]byte{}, nil, errors.Wrap(err, "could not check if block is canonical")
```
I think this is more idiomatic for our codebase.

Reply:
ping @kasey